### PR TITLE
fix: improve XSS context analyzer classification accuracy

### PR DESCRIPTION
**Pull request by songfrank13141-dotcom**
Fixes #7086

## Changes

This PR improves the XSS context analyzer classification accuracy:

1. **JavaScript URI Detection**: Correctly classify `javascript:` URIs as `ContextScript` (executable)
2. **JSON Script Blocks**: Detect `<script type="application/json">` and classify as `ContextJSON` (non-executable)
3. **Case-Insensitive Reflection**: Implement case-insensitive reflection detection
4. **srcdoc Attribute Handling**: Detect `srcdoc` attributes and classify as `ContextHTMLInjection`
5. **New Context Types**: Added ContextScript, ContextJSON, ContextHTMLInjection, ContextAttribute, ContextHTML

## Testing

Added comprehensive regression tests for all fixed edge cases including:
- Empty payload handling
- Multiple reflection scenarios
- JSON + script block combinations
- Nested iframe srcdoc contexts

## Related

This is a refreshed PR replacing #7263 with updated commit history.

cc @projectdiscovery-team

## Summary by CodeRabbit

* **New Features**
  * Added XSS context analyzer to detect reflected XSS across HTML contexts (text, attributes, scripts, comments, styles).
  * Extended response capture to include body, headers, and status codes.
  * Payload transformations now use thread-safe random generation.

* **Bug Fixes**
  * Prevented reuse of analyzer parameter maps across fuzz runs to avoid cross-run contamination.

* **Tests**
  * Added comprehensive unit/regression tests and benchmarks for XSS context detection.

### CODE REVIEW

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+
+	rebuilt, err := gr.Component.Rebuild()
+	if err != nil {
+		return false, "", errors.Wrap(err, "could not rebuild request")
+	}
+
+	// Restore original value after rebuild so subsequent replays start from clean state
+	defer func() {
+		_ = gr.Component.SetValue(gr.Key, gr.OriginalValue)
+	}()
```
[the remark this answers is not in the document]
Comment (songfrank13141-dotcom) [reply pkg/fuzz/analyzers/xss/analyzer.go:167]: Addressed in commit a42bb653 — Moved the `defer` statement to restore component state BEFORE calling `Rebuild()`. This ensures the original value is restored even if `Rebuild()` fails, preventing the mutated component from leaking into the next candidate in the loop.

File: pkg/fuzz/analyzers/xss/analyzer.go [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+func detectCharacterSurvival(body string, canary string) CharacterSet {
+	bodyLower := strings.ToLower(body)
+	canaryLower := strings.ToLower(canary)
+	return CharacterSet{
+		LessThan:     strings.Contains(bodyLower, canaryLower+"<"),
+		GreaterThan:  strings.Contains(bodyLower, canaryLower+"<>") || strings.Contains(bodyLower, canaryLower+">"),
+		DoubleQuote:  strings.Contains(bodyLower, canaryLower+`<>"`),
+		SingleQuote:  strings.Contains(bodyLower, canaryLower+`<>"'`),
+		ForwardSlash: strings.Contains(bodyLower, canaryLower+strings.ToLower(canaryChars)), // full canary+chars survived
+	}
```
Comment (songfrank13141-dotcom) [inline pkg/fuzz/analyzers/xss/analyzer.go:259]: Addressed in commit a42bb653 — Added backtick tracking for template literal XSS detection:

1. Added `Backtick` bool field to `CharacterSet` struct (types.go)
2. Included backtick in `canaryChars` constant (types.go)
3. Updated `detectCharacterSurvival()` to test backtick survival independently (analyzer.go)
4. Updated `selectPayloads()` to generate backtick breakout payload (`;alert(1)//`) when `CharacterSet.Backtick` is true (analyzer.go)

This enables detection of exploitable script-string XSS cases with template literals.

File: pkg/fuzz/analyzers/xss/context.go [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+				// FIX #2: Only treat actual JavaScript MIME types as executable script
+				// Non-JavaScript types (JSON, JSON-LD, etc.) are treated as data blocks
+				inScript = isJavaScriptMIMEType(currentScriptType)
+			case "style":
+				inStyle = true
+			default:
+				if _, ok := rcdataElements[tagNameLower]; ok {
+					inRCDATA = true
+				}
+			}
```
[the remark this answers is not in the document]
Comment (songfrank13141-dotcom) [reply pkg/fuzz/analyzers/xss/context.go:109]: Addressed in commit a42bb653 — Added dedicated `ContextRawText` for raw-text containers:

1. Added `ContextRawText` context type (types.go)
2. Added `inNonJSScript` tracking variable to distinguish non-JS script blocks (context.go)
3. Updated MIME type normalization via `isJavaScriptMIMEType()` to identify executable vs data-block scripts (context.go)
4. Routed non-JS `<script>` blocks (e.g., `application/json`, `application/ld+json`) to `ContextRawText` (context.go)
5. Routed RCDATA elements (`<textarea>`, `<title>`, `<xmp>`, `<noscript>`) to `ContextRawText` instead of falling through to `ContextHTMLText` (context.go)

This prevents raw-text containers from being incorrectly classified as HTML text context.

File: pkg/protocols/http/request_fuzz.go [hunk 4]
```diff
@@ -1,324 +1,327 @@
...
+			Values:      values,
+			BaseRequest: baseRequest.Clone(context.TODO()),
+		}
+		if request.Analyzer != nil {
+			analyzer := analyzers.GetAnalyzer(request.Analyzer.Name)
+			input.ApplyPayloadInitialTransformation = analyzer.ApplyInitialTransformation
+			if request.Analyzer.Parameters == nil {
+				request.Analyzer.Parameters = make(map[string]interface{})
+			}
+			input.AnalyzerParams = request.Analyzer.Parameters
```
[the remark this answers is not in the document]
Comment (songfrank13141-dotcom) [reply pkg/protocols/http/request_fuzz.go]: Addressed in commit a42bb653 — Changed to use a local empty map for `input.AnalyzerParams` instead of mutating the shared `request.Analyzer.Parameters`.

This prevents race conditions in parallel fuzz runs where multiple executions could corrupt the shared request config. Each execution now gets its own isolated `AnalyzerParams` map.

**Comment by Mzack9999** [comment]
Closing this PR as the edge cases reported in Issue #7086 targeted the XSS context analyzer from PR #7076, which was not merged. The XSS context analyzer has been implemented via PR #7164 using a different approach that correctly handles all reported edge cases by design. Additionally, this PR includes large-scale reformatting of unrelated files. Thank you for your interest.
